### DISCUSSION TITLE
Retrieving Historical Data

### DISCUSSION THREAD

**Discussion by Craven-Pirate** [devfinwiz/Fin-Maestro-Kin#38] [Q&A] [answered]
Hi, I am looking to integrate this in python to obtain historical data of tickers and indices? Can someone please guide me on how to do this? Also, are F&O supported?

**Comment by devfinwiz** [comment, marked as answer]
Hi @Craven-Pirate ,
You can integrate this project into your python scripts easily. Kindly refer to the example below to access https://fin-maestro-kin.apidog.io/api-6701893 endpoint responsible for obtaining index's historical data:

```
import requests

# Base URL of your running server
BASE_URL = "http://localhost:8000"

# Endpoint for fetching historical index data
endpoint = "/nseindices/history"

# Parameters for the GET request
params = {
    "symbol": "NIFTY 50",
    "start_date": "10-01-2024",
    "end_date": "12-01-2024"
}

# Sending the GET request with parameters
response = requests.get(f"{BASE_URL}{endpoint}", params=params)

# Checking if the request was successful
if response.status_code == 200:
    data = response.json()
    
    # Access the 'index_historical_data' list
    historical_data = data.get("index_historical_data", [])
    
    print("Historical Data:")
    for record in historical_data:
        print(record)
else:
    print(f"Failed to retrieve data: {response.status_code}")
```

The above code will result in this output on your terminal:
```
Historical Data:
{'index_name': 'NIFTY 50', 'open_value': 21529.3, 'high_value': 21641.85, 'close_value': 21618.7, 'low_value': 21448.65, 'timestamp': '10-JAN-2024', 'traded_quantity': 216991926, 'turnover': 20958.96}
{'index_name': 'NIFTY 50', 'open_value': 21688, 'high_value': 21726.5, 'close_value': 21647.2, 'low_value': 21593.75, 'timestamp': '11-JAN-2024', 'traded_quantity': 212453866, 'turnover': 23492.12}
{'index_name': 'NIFTY 50', 'open_value': 21773.55, 'high_value': 21928.25, 'close_value': 21894.55, 'low_value': 21715.15, 'timestamp': '12-JAN-2024', 'traded_quantity': 294678459, 'turnover': 30565.5}
```

Prerequisites:
1. Make sure you have Docker installed and set up on your machine.
2. Make sure the docker container is up and running before trying to execute the above code or accessing any endpoints. 
Reference: https://github.com/devfinwiz/Fin-Maestro-Kin#running-fin-maestro-kin-in-a-docker-container


**Note: Also, we'll soon have support for F&O as well in upcoming months. I am currently busy with something else and hence, this project's further development is on hold. We'll get back at Fin-Maestro-Kin's development starting October in full flow.**

**Reply by Craven-Pirate** [reply]
Thanks a lot. This really helped. It's working like a charm. Just wanted to know a few things about the following:
-Can we get live market data too? Is there an endpoint for that?
-is it possible to get order book data if live market data is integrated? Streaming limit order book data could be particularly useful to detect inefficiencies.
-is it possible to get OI, LTP and other things for F&O data on a daily basis?

**Reply by devfinwiz** [reply]
Glad, please mark the above response as an answer (if it was clear and useful) so it helps other developers too. (Hit 'Mark as answer').

1. Currently, there is no endpoint to gather live market data. We'll try to get that feature once we resume development in October. 
2. I did not understand the question. Order book as in company's order book to get revenue visibility or market depth to check open orders of other traders/investors?
3. Yes, we'll have an endpoint for this as well soon once we resume development. 

Enhancement to include derivatives data required, point noted!

**Reply by devfinwiz** [reply]
Also, here's the link to documentation to view all the available endpoints at one place:
https://fin-maestro-kin.apidog.io/

**Reply by Craven-Pirate** [reply]
Yes, was talking about market depth where new orders are livestreamed on the NSE website. Would be really useful if that could be integrated. bid ask spread could be captured.

**Reply by devfinwiz** [reply]
Understood. I will have it as an action item in my to do list.
